server/src/main/java/org/opensearch/common/metrics/MetricsTracker.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.metrics;
+
+import com.sun.management.ThreadMXBean;
+import org.opensearch.common.SuppressForbidden;
+
+import java.lang.management.ManagementFactory;
+
+/*
+    Simple tracker for CPU consumption and memory allocated by the current thread.
+    In the current model, it is important for updateMetrics to be called from the thread that created the object
+
+    This will evolve to support frequent updates and directly update task with resource usage
+*/
+@SuppressForbidden(reason = "ThreadMXBean enables tracking resource consumption by a thread. "
+    + "It is platform dependent and i am not aware of an alternate mechanism to extract this info")
+public class MetricsTracker {
+
+    static ThreadMXBean threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+    long startingAllocatedBytes;
+    long startingCPUTime;
+    long memoryAllocated;
+    long cpuTime;
+
+    /**
+     * Takes current snapshot of resource usage by thread since the creation of this object
+     */
+    public void updateMetrics() {
+        this.memoryAllocated = threadMXBean.getThreadAllocatedBytes(Thread.currentThread().getId()) - startingAllocatedBytes;
+        this.cpuTime = threadMXBean.getCurrentThreadCpuTime() - startingCPUTime;
+    }
+
```

Comment:
Most optimal way to achieve it is below. Same is being used in a similar PR #1643 
```
+    static {
+        threadMXBean = ManagementFactory.getThreadMXBean();
+        Method getBytes;
+        try {
+            getBytes = threadMXBean.getClass()
+                    .getMethod("getThreadAllocatedBytes", long[].class);
+            getBytes.setAccessible(true);
+            failMessages = Collections.emptyList();
+        } catch (NoSuchMethodException e) {
+            getBytes = null;
+        }
+        getThreadAllocatedBytes = getBytes;
+    }
```

Reply:
Is this for safely handling JVMs that may not have support for tracking or are there other reasons to switch to reflection - as currently this is under a cluster setting that is disabled by default i was originally planning to add and test on different platforms at a later point but can add it right away.

For safety purposes - I was wondering of leveraging [isThreadAllocatedMemorySupported](https://docs.oracle.com/en/java/javase/14/docs/api/jdk.management/com/sun/management/ThreadMXBean.html#isThreadAllocatedMemorySupported() ) and [isCurrentThreadCpuTimeSupported](https://docs.oracle.com/en/java/javase/14/docs/api/java.management/java/lang/management/ThreadMXBean.html#isCurrentThreadCpuTimeSupported() ) instead. Any other reasons why the reflection route could be better?

Reply:
We would need to confirm if those checks are good enough. See https://bugs.openjdk.java.net/browse/JDK-8152859